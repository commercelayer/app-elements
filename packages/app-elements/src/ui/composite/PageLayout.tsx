import type { ContainerProps } from '#ui/atoms/Container'
import { Container } from '#ui/atoms/Container'
import { PageHeading, type PageHeadingProps } from '#ui/atoms/PageHeading'
import { ScrollToTop } from '#ui/atoms/ScrollToTop'
import { Spacer } from '#ui/atoms/Spacer'
import { Overlay } from '#ui/internals/Overlay'
import { type ReactNode } from 'react'

export interface PageLayoutProps
  extends Pick<
      PageHeadingProps,
      'title' | 'description' | 'navigationButton' | 'actionButton' | 'gap'
    >,
    Pick<ContainerProps, 'minHeight'> {
  /**
   * Page content
   */
  children: ReactNode
  /**
   * When mode is `test`, it will render a `TEST DATA` Badge to inform user api is working in test mode.
   */
  mode?: 'test' | 'live'
  /**
   * Renders as overlay
   */
  overlay?: boolean
  /**
   * Optional prop to enable scroll to top behavior on location change
   */
  scrollToTop?: boolean
}

export function PageLayout({
  title,
  description,
  navigationButton,
  children,
  actionButton,
  mode,
  gap,
  minHeight,
  scrollToTop,
  overlay = false,
  ...rest
}: PageLayoutProps): JSX.Element {
  const component = (
    <>
      <PageHeading
        title={title}
        description={description}
        navigationButton={navigationButton}
        actionButton={actionButton}
        badgeLabel={mode === 'test' ? 'TEST DATA' : undefined}
        badgeVariant={mode === 'test' ? 'warning-solid' : undefined}
        gap={gap}
      />
      {children}
      {scrollToTop === true && <ScrollToTop />}
    </>
  )

  if (overlay) {
    return <Overlay backgroundColor='light'>{component}</Overlay>
  }

  return (
    <Container minHeight={minHeight} {...rest}>
      <Spacer bottom='14'>{component}</Spacer>
    </Container>
  )
}

PageLayout.displayName = 'PageLayout'
