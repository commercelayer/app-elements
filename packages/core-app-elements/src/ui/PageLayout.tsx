import { ReactNode } from 'react'
import { Container } from './Container'
import { PageHeading, PageHeadingProps } from './PageHeading'

interface PageLayoutProps extends PageHeadingProps {
  /**
   * Page content
   */
  children: ReactNode
}

export function PageLayout({
  title,
  description,
  onGoBack,
  children,
  ...rest
}: PageLayoutProps): JSX.Element {
  return (
    <Container {...rest}>
      <PageHeading
        title={title}
        description={description}
        onGoBack={onGoBack}
      />
      {children}
    </Container>
  )
}

export default PageLayout
