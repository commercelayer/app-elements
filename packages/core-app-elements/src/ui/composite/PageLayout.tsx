import { ReactNode } from 'react'
import { Container } from '#ui/atoms/Container'
import { PageHeading, PageHeadingProps } from '#ui/atoms/PageHeading'

interface PageLayoutProps
  extends Omit<PageHeadingProps, 'badgeVariant' | 'badgeLabel'> {
  /**
   * Page content
   */
  children: ReactNode
  /**
   * When `true` it will render a `TEST DATA` Badge to inform user test mode is on.
   */
  isTestMode?: boolean
}

export function PageLayout({
  title,
  description,
  onGoBack,
  children,
  actionButton,
  isTestMode,
  ...rest
}: PageLayoutProps): JSX.Element {
  return (
    <Container {...rest}>
      <PageHeading
        title={title}
        description={description}
        onGoBack={onGoBack}
        actionButton={actionButton}
        badgeLabel={isTestMode === true ? 'TEST DATA' : undefined}
        badgeVariant={isTestMode === true ? 'warning-solid' : undefined}
      />
      {children}
    </Container>
  )
}

export default PageLayout
