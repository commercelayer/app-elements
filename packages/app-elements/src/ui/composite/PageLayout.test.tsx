import { PageLayout, PageLayoutProps } from './PageLayout'
import { render, RenderResult } from '@testing-library/react'

interface SetupProps extends Omit<PageLayoutProps, 'children'> {
  id: string
}

type SetupResult = RenderResult & {
  element: HTMLElement
}

const setup = ({ id, ...rest }: SetupProps): SetupResult => {
  const utils = render(
    <PageLayout data-test-id={id} {...rest}>
      <div>Content...</div>
    </PageLayout>
  )
  const element = utils.getByTestId(id)
  return {
    element,
    ...utils
  }
}

describe('PageLayout', () => {
  test('Should be rendered', () => {
    const { element } = setup({
      title: 'Page title',
      id: 'my-page'
    })
    expect(element).toBeInTheDocument()
    expect(element).toMatchSnapshot()
  })

  test('Should render test mode badge', () => {
    const { getByText } = setup({
      title: 'Page title',
      id: 'my-page',
      mode: 'test'
    })
    expect(getByText('TEST DATA')).toBeInTheDocument()
  })
})
