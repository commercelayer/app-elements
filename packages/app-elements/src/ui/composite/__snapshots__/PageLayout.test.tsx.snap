// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageLayout > Should be rendered 1`] = `
<div
  class="container mx-auto flex flex-col px-4 md:px-0 min-h-screen"
  data-testid="my-page"
>
  <div
    class="mb-14"
  >
    <div
      class="w-full pt-5 md:pt-10 pb-6 md:pb-14"
    >
      <h1
        class="font-semibold text-2xl md:text-title leading-title break-words"
      >
        Page title
      </h1>
      <div
        class="text-gray-500 leading-6 mt-2"
      />
    </div>
    <div>
      Content...
    </div>
  </div>
</div>
`;
